Fixing search by filter string for params

mparkhe opened the pull request:
The issue was handling the extra set of quotes in parser.

```
"'a'" != "a"
```

Added lotsa unit tests for both `FileStore` and `SqlAlchemyStore`

Commit:
Fixing params in filter string + lotsa test cases

Commit:
Merge branch 'master' of github.com:mlflow/mlflow into search-fixing_params

andrewmchen commented on mlflow/utils/search_utils.py:
```diff
@@ -74,8 +73,10 @@ def _get_identifier(cls, identifier):
+        elif token.ttype in cls.NUMERIC_VALUE_TYPES:
             return {"value": token.value}
+        elif token.ttype in cls.STRING_VALUE_TYPES:
+            return {"value": token.value.strip("'")}  # strip quotes
```
Is this robust to something like `params.apple = '\\''`

mparkhe replied:
Legit. Lemme fix this in a few ways

- Based on identifier type (param vs metric i.e.) assert a particular type of RHS
-- params => string. detect type of quotes and trim them manually
-- metrics => numerics. cast the appropriately here
- Better error handling for wrong type of arguments

Commit:
reverting new tests added for file and sql store

Commit:
fixing test to remove quotes

Commit:
specific unit tests for param values, quoted string, white spaces (or not)

Commit:
Refactor
- Fixing ling
- Adding appropriate error_code
- asserts => MlflowException

Commit:
lint

Commit:
Merge branch 'master' of github.com:mlflow/mlflow into search-fixing_params

andrewmchen commented on mlflow/utils/search_utils.py:
```diff
@@ -37,19 +36,41 @@ def filter_string(self):
+        return string_value[1:-1]
+
+    @classmethod
+    def _is_quoted(cls, value, pattern):
```
What happens if `value` is a single character?

mparkhe replied:
It should be fine. Since we call `_is_quoted` before trimming that checks for presence of both starting and ending quote and trimming an empty string should be ok. However, I will add a `len(value) >= 2` check

andrewmchen commented on mlflow/utils/search_utils.py:
```diff
@@ -37,19 +36,41 @@ def filter_string(self):
+        """Argument value is of type string and expected to have quotes."""
+        if cls._is_quoted(value, "'") or cls._is_quoted(value, '"'):
+            return cls._trim_ends(value)
+        raise MlflowException("Parameter value is either not quoted or unidentified quote "
```
Can we say what the valid quote types are in this error message?

mparkhe replied:
👍

andrewmchen commented on mlflow/utils/search_utils.py:
```diff
@@ -71,22 +92,55 @@ def _get_identifier(cls, identifier):
+                                                                          len(tokens)),
+                                  error_code=INVALID_PARAMETER_VALUE)
+        if not isinstance(tokens[0], Identifier):
+            raise MlflowException("{}. Expected 'Identifier' found '{}'".format(base_error_string,
```
Is it possible to put the token values in these error messages?

mparkhe replied:
@andrewmchen you are a good code reviewer.

I had a tough time writing tests for these. Everything was getting caught by "needs to be a comparison clause" check. But, now that I think again--I could write test like `1 == 1` or `"LR" == params.model`

mparkhe replied:
They are included.

andrewmchen commented on mlflow/utils/search_utils.py:
```diff
@@ -71,22 +92,55 @@ def _get_identifier(cls, identifier):
+            return token.value
         else:
-            return {}
+            if identifier_type != cls._PARAM_IDENTIFIER:
```
nit: To make this function have less levels of nested branching, you could do
```
if type == METRIC:
  blah
else if type == PARAM:
  blash
else:
  raise BLAH

andrewmchen commented on mlflow/utils/search_utils.py:
```diff
@@ -71,22 +92,55 @@ def _get_identifier(cls, identifier):
+    def _validate_comparison(cls, tokens):
+        base_error_string = "Invalid comparison clause"
+        if len(tokens) != 3:
+            raise MlflowException("{}. Expected 3 tokens found {}".format(base_error_string,
```
Looks like this isn't covered in our unit tests, could we add one for this branch?

mparkhe replied:
@andrewmchen : great code review.

I had a difficult time writing test for these because anything I thought of was caught by "this needs to be a comparison clause" -- but i knew we needed this. Now that I think again, I can put in some test like `1 = 1` or `"LR" = params.model` both legit SQL syntax

Commit:
addressing review comments

Commit:
Merge branch 'master' of github.com:mlflow/mlflow into search-fixing_params

Commit:
lint

dbczumar reviewed (changes requested):
@mparkhe Awesome! This seems to have fixed the `INTERNAL_ERROR` issues associated with string quoting. Just left a nit and one other comment about an exception constructor error.

There's something else that I encountered and can't seem to figure out / may be unsupported (probably outside of the scope of this PR): 

It seems that we allow users to log params and metrics containing spaces, periods, slashes, and hyphens. However, I can't seem to figure out how to search for them.

For example, I executed the following code to log some metrics and params containing these characters:

```
import mlflow
mlflow.log_param('cat-dog', "1")
mlflow.log_param('cat.dog', "1")
mlflow.log_param('cat-dog', "2")
mlflow.log_metric('cat-dog.metric', 3)
```

If I naively attempt to query one of these parameters or metrics, I encounter errors like "invalid parameter string" or "Invalid comparison clause." For example:

1. Typing `params.cat-dog = "2"` into the search UI produces the following error text: `Invalid comparison clause. Expected 'Identifier' found 'params.cat-dog'`.

2. Typing `params.cat.dog = "2"` produces the following error text: `Invalid filter string 'params.cat.dog'. Filter comparison is expected as 'metric.<key> <comparator> <value>' or'params.<key> <comparator> <value>'.`.

3. Typing `params.cat dog = "2"` into the search UI produces the following error text: `Invalid clause(s) in filter string: 'params.cat'`

The same issues occur for similarly-named metrics. Note that these queries also fail on master (no regressions afaik). Am I supposed to wrap the parameter name in quotes or backticks? I haven't had any luck trying those approaches either. If this is functionality is unsupported, are we aware of it / do we document the limitation?

mparkhe commented on mlflow/utils/search_utils.py:
```diff
@@ -37,19 +36,41 @@ def filter_string(self):
+    def _is_quoted(cls, value, pattern):
+        if (value.startswith(pattern) and not value.endswith(pattern)) or \
+                (not value.startswith(pattern) and value.endswith(pattern)):
+            raise MlflowException("Mismatched quote types in argument {}" % value,
```
Great catch! I cannot make up my mind whether `%..` or `{} + format` is a better way to stringyfying messages.

andrewmchen replied:
The "Python" way of doing it is `.format()`. https://stackoverflow.com/a/12382738

mparkhe replied:
let's python!

mparkhe replied:
> The tests still seem to pass, which means that we aren't testing this code path. Can we add a relevant test?

Looks like this is caught up ahead -- because a mismatched quote is not valid syntax and `sqlparse` catches that. So yeah! Removing this code, 'cuz deleting code is so much better than writing test cases :)

Commit:
addressing comments. quoted keys + tests

Commit:
Merge branch 'master' of github.com:mlflow/mlflow into search-fixing_params

mparkhe commented:
> 1. Typing `params.cat-dog = "2"` into the search UI produces the following error text: `Invalid comparison clause. Expected 'Identifier' found 'params.cat-dog'`.

addressed. this needs to be `params."cat-dog" = "2"` 

> 2. Typing `params.cat.dog = "2"` produces the following error text: `Invalid filter string 'params.cat.dog'. Filter comparison is expected as 'metric.<key> <comparator> <value>' or'params.<key> <comparator> <value>'.`.

This needs to be fixed. Issue with handling period.

> 3. Typing `params.cat dog = "2"` into the search UI produces the following error text: `Invalid clause(s) in filter string: 'params.cat'`

addressed. this needs to be `params."cat dog" = "2"`

mparkhe commented on mlflow/utils/search_utils.py:
```diff
@@ -56,13 +52,19 @@ def _trim_backticks(cls, entity_type):
+    def _strip_quotes(cls, value, expect_quoted_value=False):
+        """
+        Argument values is of type string and expected to have quotes.
+        Keys with special characters are required to quoted a well.
```
@mparkhe : TODO 

```
        Keys with special characters are required to be quoted as well.
```

dbczumar reviewed (approved):
LGTM! Thanks @mparkhe

Commit:
Changing comment
